src/ralph/networks/models/networks.py

```diff
@@ -0,0 +1,416 @@
+from ralph.lib.mixins.models import (
+    AdminAbsoluteUrlMixin,
+    LastSeenMixin,
+    NamedMixin,
+    TimeStampMixin
+)
+from ralph.networks.fields import IPNetwork
+from ralph.networks.models.choices import IPAddressStatus
+
+
+class NetworkKind(NamedMixin, models.Model):
+    class Meta:
+        verbose_name = _('network kind')
+        verbose_name_plural = _('network kinds')
+        ordering = ('name',)
+
+    def __str__(self):
+        return self.name
+
+
+class NetworkEnvironment(NamedMixin):
+    data_center = models.ForeignKey(
+        'data_center.DataCenter',
+        verbose_name=_('data center')
+    )
+    queue = models.ForeignKey(
+        'DiscoveryQueue',
+        verbose_name=_('discovery queue'),
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+    )
+    hosts_naming_template = models.CharField(
+        verbose_name=_('hosts naming template'),
+        max_length=30,
+        help_text=_(
+            'E.g. h<200,299>.dc|h<400,499>.dc will produce: h200.dc '
+            'h201.dc ... h299.dc h400.dc h401.dc'
+        ),
+    )
+    next_server = models.CharField(
+        verbose_name=_('next server'),
+        max_length=32,
+        blank=True,
+        default='',
+        help_text=_('The address for a TFTP server for DHCP.'),
+    )
+    domain = models.CharField(
+        verbose_name=_('domain'),
+        max_length=255,
+        blank=True,
+        null=True,
+    )
+    remarks = models.TextField(
+        verbose_name=_('remarks'),
+        help_text=_('Additional information.'),
+        blank=True,
+        null=True,
+    )
+
+    def __str__(self):
+        return self.name
+
+    class Meta:
+        ordering = ('name',)
+
+
+class NetworkMixin(object):
+    _parent_attr = None
+
+    def _assign_parent(self):
+        parent = getattr(self, self._parent_attr)
+        if not parent or self.pk:
+            setattr(self, self._parent_attr, self.get_network())
+
+    def search_networks(self):
+        raise NotImplementedError()
+
+    def get_network(self):
+        network = None
+        networks = self.search_networks()
+        if networks:
+            network = networks[0]
+        return network
+
+
+class Network(
+    AdminAbsoluteUrlMixin,
+    NamedMixin,
+    TimeStampMixin,
+    NetworkMixin,
+    MPTTModel,
+    models.Model
+):
+    """
+    Class for networks.
+    """
+    _parent_attr = 'parent'
+    parent = TreeForeignKey(
+        'self',
+        null=True,
+        blank=True,
+        related_name='children',
+        db_index=True,
+        editable=False,
+    )
+    address = IPNetwork(
+        verbose_name=_('network address'),
+        help_text=_('Presented as string (e.g. 192.168.0.0/24)'),
+    )
+    remarks = models.TextField(
+        verbose_name=_('remarks'),
+        help_text=_('Additional information.'),
+        blank=True,
+        default='',
+    )
+    terminators = models.ManyToManyField(
```

Comment:
afair we should link it with `BaseObject`, not separated model with char field. look at `BaseObjectForeignKey`